README
-*- mode: text -*-

XSLT-processor -- A JAVASCRIPT XSLT PROCESSOR LIBRARY


INTRODUCTION

 XSLT-processor is an implementation of XSLT in JavaScript. Because XSLT uses
 XPath, it is also an implementation of XPath that can be used
 independently of XSLT. This implementation has the advantange that it
 makes XSLT uniformly available on more browsers than natively provide
 it, and that it can be extended to still more browsers if
 necessary.

 XSLT-processor is interesting for developers who strive aggressively for
 cross browser compatibility of their advanced web applications.

 This implementation of XSLT operates at the DOM level on its input
 documents. It internally uses a DOM implementation to create the
 output document, but usually returns the output document as text
 stream. The DOM to construct the output document can be supplied by
 the application, or else an internal minimal DOM implementation is
 used. This DOM comes with a minimal XML parser that can be used to
 generate a suitable DOM representation of the input documents if they
 are present as text.


TESTS AND USAGE EXAMPLES

 The files ending in _unittest.html are jsunit test pages. See
 <http://www.jsunit.net/>.

 The files xslt.html and xpath.html in the directory test are
 interactive tests. They can be run directly from the file system; no
 HTTP server is needed.

 Both unit tests and interactive tests demonstrate the use of the
 library functions. There is not much more documentation so far.

CONFORMANCE

 A few features that are required by the XSLT and XPath standards were
 left out. They are marked in the source code using alert()
 statements. See xslt.js.

 The DOM implementation is minimal so as to support the XSLT
 processing, and not intended to be complete.

 The implementation is all agnostic about namespaces. It just expects
 XSLT elements to have tags that carry the xsl: prefix, but we
 disregard all namespace declaration for them.

 There are a few nonstandard XPath functions. Grep xpath.js for "ext-"
 to see their definitions.
 

REFERENCES

 [XPATH] XPath Specification
     <http://www.w3.org/TR/1999/REC-xpath-19991116>.
 [XSLT] XSLT Specification
     <http://www.w3.org/TR/1999/REC-xslt-19991116>.
 [DOM] W3C DOM Level 3 Core Specification
     <http://www.w3.org/TR/2004/REC-DOM-Level-3-Core-20040407/>.
 [ECMA] ECMAScript Language Specification
     <http://www.ecma-international.org/publications/standards/Ecma-262.htm>.


ChangeLog
2018-03-20  Johannes Wilm <johannes@fiduswriter.org>

	* Version 0.9.0
	* Changed code to ES2015+ style
	* Fixed a bug where xml parsing choked on < and > inside of quotation marks.
	* Renamed to xslt-processor

2007-11-16  Google Inc.  <opensource@google.com>

	* Version 0.8
	* Released as http://ajaxslt.googlecode.com/svn/tags/release-0-8/

	* Fixed a bug in parsing XPaths that end with "//qname" (Issue 17)

	* Added feature to optionally allow case-insensitive node name
	comparisons; this is useful when using XPaths on HTML, where
	node names are not consistent across browsers.

	* Improved performance by relying on getElementsByTagName where
	possible

	* Workaround IE bug where "javascript:" href attribute is URL
	encoded.  (Issue 19)

2006-12-28  Google Inc.  <opensource@google.com>

	* Version 0.7
	* Released as http://ajaxslt.googlecode.com/svn/tags/release-0-7/

	* Fixed a bug that semicolons are dropped by the XML parser when a
	text nodes also contains an entity.

	* Fixed a bug that xsl:variable definitions with a node set value
	encountered at the root level of the input document would throw an
	exception.

	* Fixed a bug that XPath expression @* always evaluated to the
	empty node set.

	* Fixed a bug that xsl:copy would copy only attribute and element
	nodes.

	* Fixed a bug that if xsl:apply-templates matches multiple
	templates, the output is sorted according to the order of the
	matching templates, and not according to the sort order defined
	for the selected node set to which templates are applied.

	* Added unittests for all fixed bugs.

	* Added wrapper function xmlOwnerDocument() to uniformly access
	the document on both document nodes and other nodes and use it
	throughout the xslt processor.

2006-12-14  Google Inc.  <opensource@google.com>

	* Version 0.6
	* Released as http://ajaxslt.googlecode.com/svn/tags/release-0-6/
	* Fixes infinite loops in evaluation of XPath axes "ancestor",
	  "ancestor-or-self", "preceding-sibling", "following-sibling".
	* Fixes evaluation of XPath axes "preceding", "following".
	* Added unittests for both.
	* Fixed xmlEscape*() functions to escape *all* markup characters
	  not just the first.
	* Fixed xsl:copy-of to also copy CDATA and COMMENT nodes.

2006-09-10  Google Inc. <opensource@google.com>

	* Version 0.5
	* Released on http://code.google.com/hosting/
	* General changes:
	  - remove all uses of for-in iteration
	  - rename misc.js to util.js
	  - log window is now in simplelog.js
	* XPath changes:
	  - fixed id() function
	  - fixed UnionExpr::evaluate()
	  - added support for Unicode chracters
	* XSLT changes:
	  - fixed xsl:sort in xsl:for-each (again)
	* DOM changes:
	  - added a few methods
	* XML parser changes:
	  - parses CDATA sections
	  - parses comments
	  - parses XML declaration
	  - parses Unicode XML markup
	* Test changes:
	  - added several jsunit tests

2005-10-19  Google Inc. <opensource@google.com>

	* Version 0.4
	* XPath changes:
	  - Optimize parsing of very common and simple expressions.
	  - Fix use of XPath operator names -- div, mod, and, or --
	    as node names in abbreviated step expressions.
	  - Fix root node -- it is now set to ownerDocument.
	* XSLT changes:
	  - Fix xsl:sort in xsl:for-each.
	* DOM changes:
	  - Add replaceChild(), insertBefore(), removeChild().
	    These are still not needed in XSLT processing, but
	    in another client of the DOM implementation.
	  - DOM nodes are recycled instead of garbage collected,
	    in order to improve performance in some browsers.
	* Test changes:
	  - Add many more test cases to the XPath tests.
	  - Add a note mentioning jsunit in the README.
	  - Add a DOM unittest file.

2005-08-27  Google Inc. <opensource@google.com>

	* Version 0.3 (not released on sourceforge)
	* XPath changes:
	  - Fix implementation of the * node test.
	  - Fix implementation of the substring() function.
	  - Fix non-abbreviated axis names.
	  - Fix filter expressions.
	* XSLT changes:
	  - Fix xsl:sort.
	* DOM changes:
	  - Avoid using String.split() that breaks in IE.
	  - Coerce nodeType to number and nodeName and nodeValue to string.
	  - Fix SGML entity replacement of single quotes in attribute values.
	* Test changes:
	  - Fix end tags of script elements in test HTML files.
	  - Add jsunit tests for xpath.js.


2005-06-29  Google Inc. <opensource@google.com>

	* Version 0.2
	* Add more missing code
	  - XML parser and simple DOM implementation in dom.js
	  - miscellaneous functions in misc.js.
	* Add simple test pages that serve as examples.
	  - test-xpath.html tests and demonstrates the XPath parser.
	  - test-xslt.html tests and demonstrates the XSLT processor.
	  - output methods for debugging of XPath expressions added
	    in xpathdebug.js.
	* Some additions and corrections in README and TODO

src/xslt.js
// Copyright 2018 Johannes Wilm
// Copyright 2005 Google Inc.
// All Rights Reserved
//
//
// An XSL-T processor written in JavaScript. The implementation is NOT
// complete; some xsl element are left out.
//
// References:
//
// [XSLT] XSL-T Specification
// <http://www.w3.org/TR/1999/REC-xslt-19991116>.
//
// [ECMA] ECMAScript Language Specification
// <http://www.ecma-international.org/publications/standards/Ecma-262.htm>.
//
// The XSL processor API has one entry point, the function
// xsltProcessContext(). It receives as arguments the starting point in the
// input document as an XPath expression context, the DOM root node of
// the XSL-T stylesheet, and a DOM node that receives the output.
//
// NOTE: Actually, XSL-T processing according to the specification is
// defined as operation on text documents, not as operation on DOM
// trees. So, strictly speaking, this implementation is not an XSL-T
// processor, but the processing engine that needs to be complemented
// by an XML parser and serializer in order to be complete. Those two
// are found in the file xml.js.
//
//
// TODO(mesch): add jsdoc comments. Use more coherent naming. Finish
// remaining XSLT features.
//
//
// Original author: Steffen Meschkat <mesch@google.com>
import {
    xmlValue,
    xmlText,
    xmlOwnerDocument,
    domGetAttribute,
    domSetAttribute,
    domAppendChild,
    domCreateTextNode,
    domCreateElement,
    domCreateCDATASection,
    domCreateComment,
    domCreateDocumentFragment
} from "./util.js"
import {
    xpathParse,
    ExprContext,
    StringValue,
    NodeSetValue,
    xpathSort,
    xpathEval
} from "./xpath.js"
import {
    XDocument,
    xmlResolveEntities,
    DOM_DOCUMENT_NODE,
    DOM_DOCUMENT_FRAGMENT_NODE,
    DOM_CDATA_SECTION_NODE,
    DOM_ELEMENT_NODE,
    DOM_TEXT_NODE,
    DOM_COMMENT_NODE,
    DOM_ATTRIBUTE_NODE
} from "./dom.js"

// The exported entry point of the XSL-T processor, as explained
// above.
//
// @param xmlDoc The input document root, as DOM node.
// @param template The stylesheet document root, as DOM node.
// @return the processed document, as XML text in a string.
export function xsltProcess(xmlDoc, stylesheet) {
    const output = domCreateDocumentFragment(new XDocument);
    xsltProcessContext(new ExprContext(xmlDoc), stylesheet, output);
    const ret = xmlText(output);
    return ret;
}

// The main entry point of the XSL-T processor, as explained above.
//
// @param input The input document root, as XPath ExprContext.
// @param template The stylesheet document root, as DOM node.
// @param the root of the generated output, as DOM node.

function xsltProcessContext(input, template, output) {
    const outputDocument = xmlOwnerDocument(output);

    const nodename = template.nodeName.split(/:/);
    if (nodename.length == 1 || nodename[0] != 'xsl') {
        xsltPassThrough(input, template, output, outputDocument);
    } else {
        switch (nodename[1]) {
            case 'apply-imports':
                alert(`not implemented: ${nodename[1]}`);
                break;

            case 'apply-templates':
                var select = xmlGetAttribute(template, 'select');
                var nodes;
                if (select) {
                    nodes = xpathEval(select, input).nodeSetValue();
                } else {
                    nodes = input.node.childNodes;
                }

                const sortContext = input.clone(nodes[0], 0, nodes);
                xsltWithParam(sortContext, template);
                xsltSort(sortContext, template);

                const mode = xmlGetAttribute(template, 'mode');
                var top = template.ownerDocument.documentElement;
                const templates = [];
                for (let i = 0; i < top.childNodes.length; ++i) {
                    let c = top.childNodes[i];
                    if (c.nodeType == DOM_ELEMENT_NODE &&
                        c.nodeName == 'xsl:template' &&
                        (!mode || c.getAttribute('mode') == mode)
                    ) {
                        templates.push(c);
                    }
                }
                for (let j = 0; j < sortContext.contextSize(); ++j) {
                    const nj = sortContext.nodelist[j];
                    for (let i = 0; i < templates.length; ++i) {
                        xsltProcessContext(sortContext.clone(nj, j), templates[i], output);
                    }
                }
                break;

            case 'attribute':
                var nameexpr = xmlGetAttribute(template, 'name');
                var name = xsltAttributeValue(nameexpr, input);
                var node = domCreateDocumentFragment(outputDocument);
                xsltChildNodes(input, template, node);
                var value = xmlValue(node);
                domSetAttribute(output, name, value);
                break;

            case 'attribute-set':
                alert(`not implemented: ${nodename[1]}`);
                break;

            case 'call-template':
                var name = xmlGetAttribute(template, 'name');
                var top = template.ownerDocument.documentElement;

                const paramContext = input.clone();
                xsltWithParam(paramContext, template);

                for (var i = 0; i < top.childNodes.length; ++i) {
                    var c = top.childNodes[i];
                    if (c.nodeType == DOM_ELEMENT_NODE &&
                        c.nodeName == 'xsl:template' &&
                        domGetAttribute(c, 'name') == name) {
                        xsltChildNodes(paramContext, c, output);
                        break;
                    }
                }
                break;

            case 'choose':
                xsltChoose(input, template, output);
                break;

            case 'comment':
                var node = domCreateDocumentFragment(outputDocument);
                xsltChildNodes(input, template, node);
                const commentData = xmlValue(node);
                const commentNode = domCreateComment(outputDocument, commentData);
                output.appendChild(commentNode);
                break;

            case 'copy':
                var node = xsltCopy(output, input.node, outputDocument);
                if (node) {
                    xsltChildNodes(input, template, node);
                }
                break;

            case 'copy-of':
                var select = xmlGetAttribute(template, 'select');
                var value = xpathEval(select, input);
                if (value.type == 'node-set') {
                    var nodes = value.nodeSetValue();
                    for (var i = 0; i < nodes.length; ++i) {
                        xsltCopyOf(output, nodes[i], outputDocument);
                    }

                } else {
                    var node = domCreateTextNode(outputDocument, value.stringValue());
                    domAppendChild(output, node);
                }
                break;

            case 'decimal-format':
                alert(`not implemented: ${nodename[1]}`);
                break;

            case 'element':
                var nameexpr = xmlGetAttribute(template, 'name');
                var name = xsltAttributeValue(nameexpr, input);
                var node = domCreateElement(outputDocument, name);
                domAppendChild(output, node);
                xsltChildNodes(input, template, node);
                break;

            case 'fallback':
                alert(`not implemented: ${nodename[1]}`);
                break;

            case 'for-each':
                xsltForEach(input, template, output);
                break;

            case 'if':
                const test = xmlGetAttribute(template, 'test');
                if (xpathEval(test, input).booleanValue()) {
                    xsltChildNodes(input, template, output);
                }
                break;

            case 'import':
                alert(`not implemented: ${nodename[1]}`);
                break;

            case 'include':
                alert(`not implemented: ${nodename[1]}`);
                break;

            case 'key':
                alert(`not implemented: ${nodename[1]}`);
                break;

            case 'message':
                alert(`not implemented: ${nodename[1]}`);
                break;

            case 'namespace-alias':
                alert(`not implemented: ${nodename[1]}`);
                break;

            case 'number':
                alert(`not implemented: ${nodename[1]}`);
                break;

            case 'otherwise':
                alert(`error if here: ${nodename[1]}`);
                break;

            case 'output':
                // Ignored. -- Since we operate on the DOM, and all further use
                // of the output of the XSL transformation is determined by the
                // browser that we run in, this parameter is not applicable to
                // this implementation.
                break;

            case 'preserve-space':
                alert(`not implemented: ${nodename[1]}`);
                break;

            case 'processing-instruction':
                alert(`not implemented: ${nodename[1]}`);
                break;

            case 'sort':
                // just ignore -- was handled by xsltSort()
                break;

            case 'strip-space':
                alert(`not implemented: ${nodename[1]}`);
                break;

            case 'stylesheet':
            case 'transform':
                xsltChildNodes(input, template, output);
                break;

            case 'template':
                const match = xmlGetAttribute(template, 'match');
                if (match && xsltMatch(match, input)) {
                    xsltChildNodes(input, template, output);
                }
                break;

            case 'text':
                const text = xmlValue(template);
                var node = domCreateTextNode(outputDocument, text);
                output.appendChild(node);
                break;

            case 'value-of':
                var select = xmlGetAttribute(template, 'select');
                var value = xpathEval(select, input).stringValue();
                var node = domCreateTextNode(outputDocument, value);
                output.appendChild(node);
                break;

            case 'param':
                xsltVariable(input, template, false);
                break;

            case 'variable':
                xsltVariable(input, template, true);
                break;

            case 'when':
                alert(`error if here: ${nodename[1]}`);
                break;

            case 'with-param':
                alert(`error if here: ${nodename[1]}`);
                break;

            default:
                alert(`error if here: ${nodename[1]}`);
                break;
        }
    }
}


// Sets parameters defined by xsl:with-param child nodes of the
// current template node, in the current input context. This happens
// before the operation specified by the current template node is
// executed.

function xsltWithParam(input, template) {
    for (const c of template.childNodes) {
        if (c.nodeType == DOM_ELEMENT_NODE && c.nodeName == 'xsl:with-param') {
            xsltVariable(input, c, true);
        }
    }
}


// Orders the current node list in the input context according to the
// sort order specified by xsl:sort child nodes of the current
// template node. This happens before the operation specified by the
// current template node is executed.
//
// TODO(mesch): case-order is not implemented.

function xsltSort(input, template) {
    const sort = [];

    for (const c of template.childNodes) {
        if (c.nodeType == DOM_ELEMENT_NODE && c.nodeName == 'xsl:sort') {
            const select = xmlGetAttribute(c, 'select');
            const expr = xpathParse(select);
            const type = xmlGetAttribute(c, 'data-type') || 'text';
            const order = xmlGetAttribute(c, 'order') || 'ascending';
            sort.push({
                expr,
                type,
                order
            });
        }
    }

    xpathSort(input, sort);
}


// Evaluates a variable or parameter and set it in the current input
// context. Implements xsl:variable, xsl:param, and xsl:with-param.
//
// @param override flag that defines if the value computed here
// overrides the one already in the input context if that is the
// case. I.e. decides if this is a default value or a local
// value. xsl:variable and xsl:with-param override; xsl:param doesn't.

function xsltVariable(input, template, override) {
    const name = xmlGetAttribute(template, 'name');
    const select = xmlGetAttribute(template, 'select');

    let value;

    if (template.childNodes.length > 0) {
        const root = domCreateDocumentFragment(template.ownerDocument);
        xsltChildNodes(input, template, root);
        value = new NodeSetValue([root]);

    } else if (select) {
        value = xpathEval(select, input);

    } else {
        value = new StringValue('');
    }

    if (override || !input.getVariable(name)) {
        input.setVariable(name, value);
    }
}


// Implements xsl:chose and its child nodes xsl:when and
// xsl:otherwise.

function xsltChoose(input, template, output) {
    for (const childNode of template.childNodes) {
        if (childNode.nodeType != DOM_ELEMENT_NODE) {
            continue;

        } else if (childNode.nodeName == 'xsl:when') {
            const test = xmlGetAttribute(childNode, 'test');
            if (xpathEval(test, input).booleanValue()) {
                xsltChildNodes(input, childNode, output);
                break;
            }

        } else if (childNode.nodeName == 'xsl:otherwise') {
            xsltChildNodes(input, childNode, output);
            break;
        }
    }
}


// Implements xsl:for-each.

function xsltForEach(input, template, output) {
    const select = xmlGetAttribute(template, 'select');
    const nodes = xpathEval(select, input).nodeSetValue();
    const sortContext = input.clone(nodes[0], 0, nodes);
    xsltSort(sortContext, template);
    for (let i = 0; i < sortContext.contextSize(); ++i) {
        const ni = sortContext.nodelist[i];
        xsltChildNodes(sortContext.clone(ni, i), template, output);
    }
}


// Traverses the template node tree. Calls the main processing
// function with the current input context for every child node of the
// current template node.

function xsltChildNodes(input, template, output) {
    // Clone input context to keep variables declared here local to the
    // siblings of the children.
    const context = input.clone();
    for (let i = 0; i < template.childNodes.length; ++i) {
        xsltProcessContext(context, template.childNodes[i], output);
    }
}


// Passes template text to the output. The current template node does
// not specify an XSL-T operation and therefore is appended to the
// output with all its attributes. Then continues traversing the
// template node tree.

function xsltPassThrough(input, template, output, outputDocument) {
    if (template.nodeType == DOM_TEXT_NODE) {
        if (xsltPassText(template)) {
            var node = domCreateTextNode(outputDocument, template.nodeValue);
            domAppendChild(output, node);
        }

    } else if (template.nodeType == DOM_ELEMENT_NODE) {
        var node = domCreateElement(outputDocument, template.nodeName);

        for (const a of template.attributes) {
            if (a) {
                const name = a.nodeName;
                const value = xsltAttributeValue(a.nodeValue, input);
                domSetAttribute(node, name, value);
            }
        }

        domAppendChild(output, node);
        xsltChildNodes(input, template, node);
    } else {
        // This applies also to the DOCUMENT_NODE of the XSL stylesheet,
        // so we don't have to treat it specially.
        xsltChildNodes(input, template, output);
    }
}

// Determines if a text node in the XSLT template document is to be
// stripped according to XSLT whitespace stipping rules.
//
// See [XSLT], section 3.4.
//
// TODO(mesch): Whitespace stripping on the input document is
// currently not implemented.

function xsltPassText(template) {
    if (!template.nodeValue.match(/^\s*$/)) {
        return true;
    }

    let element = template.parentNode;
    if (element.nodeName == 'xsl:text') {
        return true;
    }

    while (element && element.nodeType == DOM_ELEMENT_NODE) {
        const xmlspace = domGetAttribute(element, 'xml:space');
        if (xmlspace) {
            if (xmlspace == 'default') {
                return false;
            } else if (xmlspace == 'preserve') {
                return true;
            }
        }

        element = element.parentNode;
    }

    return false;
}

// Evaluates an XSL-T attribute value template. Attribute value
// templates are attributes on XSL-T elements that contain XPath
// expressions in braces {}. The XSL-T expressions are evaluated in
// the current input context.

function xsltAttributeValue(value, context) {
    const parts = value.split('{');
    if (parts.length == 1) {
        return value;
    }

    let ret = '';
    for (let i = 0; i < parts.length; ++i) {
        const rp = parts[i].split('}');
        if (rp.length != 2) {
            // first literal part of the value
            ret += parts[i];
            continue;
        }

        const val = xpathEval(rp[0], context).stringValue();
        ret += val + rp[1];
    }

    return ret;
}


// Wrapper function to access attribute values of template element
// nodes. Currently this calls xmlResolveEntities because in some DOM
// implementations the return value of node.getAttributeValue()
// contains unresolved XML entities, although the DOM spec requires
// that entity references are resolved by te DOM.
function xmlGetAttribute(node, name) {
    // TODO(mesch): This should not be necessary if the DOM is working
    // correctly. The DOM is responsible for resolving entities, not the
    // application.
    const value = domGetAttribute(node, name);
    if (value) {
        return xmlResolveEntities(value);
    } else {
        return value;
    }
};


// Implements xsl:copy-of for node-set values of the select
// expression. Recurses down the source node tree, which is part of
// the input document.
//
// @param {Node} dst the node being copied to, part of output document,
// @param {Node} src the node being copied, part in input document,
// @param {Document} dstDocument

function xsltCopyOf(dst, src, dstDocument) {
    if (src.nodeType == DOM_DOCUMENT_FRAGMENT_NODE ||
        src.nodeType == DOM_DOCUMENT_NODE) {
        for (var i = 0; i < src.childNodes.length; ++i) {
            xsltCopyOf(dst, src.childNodes[i], dstDocument);
        }
    } else {
        const node = xsltCopy(dst, src, dstDocument);
        if (node) {
            // This was an element node -- recurse to attributes and
            // children.
            for (var i = 0; i < src.attributes.length; ++i) {
                xsltCopyOf(node, src.attributes[i], dstDocument);
            }

            for (var i = 0; i < src.childNodes.length; ++i) {
                xsltCopyOf(node, src.childNodes[i], dstDocument);
            }
        }
    }
}


// Implements xsl:copy for all node types.
//
// @param {Node} dst the node being copied to, part of output document,
// @param {Node} src the node being copied, part in input document,
// @param {Document} dstDocument
// @return {Node|Null} If an element node was created, the element
// node. Otherwise null.

function xsltCopy(dst, src, dstDocument) {
    if (src.nodeType == DOM_ELEMENT_NODE) {
        var node = domCreateElement(dstDocument, src.nodeName);
        domAppendChild(dst, node);
        return node;
    }

    if (src.nodeType == DOM_TEXT_NODE) {
        var node = domCreateTextNode(dstDocument, src.nodeValue);
        domAppendChild(dst, node);

    } else if (src.nodeType == DOM_CDATA_SECTION_NODE) {
        var node = domCreateCDATASection(dstDocument, src.nodeValue);
        domAppendChild(dst, node);

    } else if (src.nodeType == DOM_COMMENT_NODE) {
        var node = domCreateComment(dstDocument, src.nodeValue);
        domAppendChild(dst, node);

    } else if (src.nodeType == DOM_ATTRIBUTE_NODE) {
        domSetAttribute(dst, src.nodeName, src.nodeValue);
    }

    return null;
}


// Evaluates an XPath expression in the current input context as a
// match (see [XSLT] section 5.2, paragraph 1).
function xsltMatch(match, context) {
    const expr = xpathParse(match);
    let ret;
    // Shortcut for the most common case.
    if (expr.steps && !expr.absolute && expr.steps.length == 1 &&
        expr.steps[0].axis == 'child' && expr.steps[0].predicate.length == 0) {
        ret = expr.steps[0].nodetest.evaluate(context).booleanValue();

    } else {

        ret = false;
        let node = context.node;

        while (!ret && node) {
            const result = expr.evaluate(context.clone(node, 0, [node])).nodeSetValue();
            for (let i = 0; i < result.length; ++i) {
                if (result[i] == context.node) {
                    ret = true;
                    break;
                }
            }
            node = node.parentNode;
        }
    }
    return ret;
}
